query-builder/src/main/java/com/datastax/oss/driver/internal/querybuilder/schema/DefaultCreateTable.java

```diff
@@ -0,0 +1,319 @@
+        tableName,
+        ifNotExists,
+        compactStorage,
+        columnsInOrder,
+        partitionKeyColumns,
+        clusteringKeyColumns,
+        staticColumns,
+        regularColumns,
+        orderings,
+        options);
+  }
+
+  @Override
+  public String asCql() {
+    StringBuilder builder = new StringBuilder();
+
+    builder.append("CREATE TABLE ");
+    if (ifNotExists) {
+      builder.append("IF NOT EXISTS ");
+    }
+
+    CqlHelper.qualify(keyspace, tableName, builder);
+
+    boolean singlePrimaryKey = partitionKeyColumns.size() == 1 && clusteringKeyColumns.size() == 0;
+
+    builder.append(" (");
+
+    boolean first = true;
+    for (Map.Entry<CqlIdentifier, DataType> column : columnsInOrder.entrySet()) {
+      if (first) {
+        first = false;
+      } else {
+        builder.append(',');
+      }
+      builder
+          .append(column.getKey().asCql(true))
+          .append(' ')
+          .append(column.getValue().asCql(true, true));
+
+      if (singlePrimaryKey && partitionKeyColumns.contains(column.getKey())) {
+        builder.append(" PRIMARY KEY");
+      } else if (staticColumns.contains(column.getKey())) {
+        builder.append(" STATIC");
+      }
+    }
+
+    if (!singlePrimaryKey) {
+      builder.append(",");
+      CqlHelper.buildPrimaryKey(partitionKeyColumns, clusteringKeyColumns, builder);
+    }
+
+    builder.append(')');
+
+    if (compactStorage || !orderings.isEmpty() || !options.isEmpty()) {
+      boolean firstOption = true;
+
+      if (compactStorage) {
+        firstOption = false;
+        builder.append(" WITH COMPACT STORAGE");
+      }
+
+      if (!orderings.isEmpty()) {
+        if (firstOption) {
+          builder.append(" WITH ");
+          firstOption = false;
+        } else {
+          builder.append(" AND ");
+        }
+        builder.append("CLUSTERING ORDER BY (");
+        boolean firstClustering = true;
+
+        for (Map.Entry<CqlIdentifier, ClusteringOrder> ordering : orderings.entrySet()) {
+          if (firstClustering) {
+            firstClustering = false;
+          } else {
+            builder.append(',');
+          }
+          builder
+              .append(ordering.getKey().asCql(true))
+              .append(' ')
+              .append(ordering.getValue().toString());
+        }
+
+        builder.append(')');
+      }
+
+      builder.append(OptionsUtils.buildOptions(options, firstOption));
+    }
+
+    return builder.toString();
+  }
+
+  @Override
+  public CreateTable withOption(String name, Object value) {
+    return new DefaultCreateTable(
+        keyspace,
+        tableName,
+        ifNotExists,
+        compactStorage,
+        columnsInOrder,
+        partitionKeyColumns,
+        clusteringKeyColumns,
+        staticColumns,
+        regularColumns,
+        orderings,
+        ImmutableCollections.append(options, name, value));
+  }
+
+  @Override
+  public Map<String, Object> getOptions() {
+    return options;
+  }
+}
```

Comment:
All implementation classes should have:

* a `toString()` implementation that calls `asCql()` (sadly we can't have it as a default method of `BuildableQuery`). Also make sure it always works, even when the query is not buildable yet.
* public getters for their internal fields (for introspection purposes).

Reply:
👍 will take care of that.